Add refresh button to unified resources view

gzdunek opened the pull request:
@roraback please take a look a the UI:

https://github.com/user-attachments/assets/a965d3e2-be72-477b-8753-140e08c0bfe1

-----------
We also [discussed refreshing the view with Cmd+R in Connect](https://gravitational.slack.com/archives/C03FJA391M3/p1721918452424409?thread_ts=1721916999.187949&cid=C03FJA391M3), but I ran into a problem with that.
On Windows/Linux the shortcut should be Ctrl+R, but that would conflict with a bash shortcut. To solve the problem, the shortcut should only work when the resources view is visible.
1. To achieve this, we would need to refactor the `keyboardShortcutService` so that it only stops event propagation when there is a handler registered for a given shortcut (currently we stop propagation for each defined shortcut).
3. The unified view in `DocumentCluster` would have to register a handler only when the document is visible:
```ts
    useKeyboardShortcuts({ 'refresh': () => props.visible ? fetch({ clear: true }) : undefined });
``` 
The only problem is that it would degrade performance on the unified view, since passing `visible` to it would cause re-rendering it when the active document changes. We could workaround this by extracting an intermediate component wrapped in `memo`, but I'm wondering if this also worth the effort ;) Or maybe there is an easier way to solve this problem that I don't see?

gzdunek committed:
Allow `useKeyBasedPagination.clear` to clear the state before fetching new items

gzdunek committed:
Add refresh button to unified resources view

avatus reviewed (approved):
neat!

roraback commented:
@gzdunek it looks good, but is it a bit wider than it is tall? Ideally these icon buttons should be square, w/minimum dimensions of 24x24 for the smallest size.

gzdunek committed:
Make the icon button square

gzdunek committed:
Remove redundant border color

gzdunek commented:
Good eye, @roraback! The button was 28x24. 
For now, I fixed manually this specific case. I talked about it with @bl-nero, he is going to address this globally.

Fixed icon: 
<img width="650" alt="image" src="https://github.com/user-attachments/assets/c888dbee-dca2-447a-a9fc-d8477ca3c32e">

ravicious commented on web/packages/shared/components/UnifiedResources/FilterPanel.tsx:
> To achieve this, we would need to refactor the `keyboardShortcutService` so that it only stops event propagation when there is a handler registered for a given shortcut (currently we stop propagation for each defined shortcut).

This is worth addressing at some point, otherwise it's going to bug us whenever there will be an opportunity to add a small shortcut like this one. OTOH, so far this is the only non-global shortcut that we wanted to add.

> The only problem is that it would degrade performance on the unified view, since passing `visible` to it would cause re-rendering it when the active document changes. We could workaround this by extracting an intermediate component wrapped in `memo`, but I'm wondering if this also worth the effort ;)

Do you remember why we use `memo` for `Resources` in the first place? IIRC it was an issue with how deps are specified within some of the effects within unified resources? Though even with that in place, I'm not 100% sure if passing `visible` to `Resources` would be a problem at all. 🤔 The way the hierarchy looks like at the moment is (IIRC):

```tsx
<MemoizedDocument doc={doc} visible={visible}>
  // …
  <MemoizedResources (a bunch of props needed for unified resources) />
```

Adding `visible` to `MemoizedResources` would be one extra rerender whenever the tab becomes active or becomes inactive. I don't remember how bad it would be.

Ultimately the problem is that `useUnifiedResourcesFetch` returns unstable functions, right? Again, I'm not sure if I remember this correctly. If it wasn't the case, I guess `useUnifiedResourcesFetch` could be moved up, but then we'd probably not need `memo` in the first place.

I'm very curious what React Compiler is going to do with this component. :D

gzdunek replied:
>Do you remember why we use memo for Resources in the first place? IIRC it was an issue with how deps are specified within some of the effects within unified resources? Though even with that in place, I'm not 100% sure if passing visible to Resources would be a problem at all. 

We use `memo` because rendering the `shared/UnifiedResources components is expensive. It is visible when quickly switching between tabs (it's not smooth, [see the original PR](https://github.com/gravitational/teleport/pull/36741#issue-2084194439)). 

>Adding visible to MemoizedResources would be one extra rerender whenever the tab becomes active or becomes inactive. I don't remember how bad it would be.

It's not as bad as before the linked PR, but it's noticeable. 

>Ultimately the problem is that useUnifiedResourcesFetch returns unstable functions, right? Again, I'm not sure if I remember this correctly. If it wasn't the case, I guess useUnifiedResourcesFetch could be moved up, but then we'd probably not need memo in the first place.

Actually, `useUnifiedResourcesFetch` returns stable functions, but we would need the intermediate component anyway. It's because we pass a custom `key` to the memoized `Resources` component (to reset it, when query params change). I think the hierarchy should look like below:
 
```ts
<teleterm.UnifiedResources>
  <MemoizedResourcesWithFetchAndKeyboardShortcut key={}>
    <MemoizedResources>
```

But at the moment I'd rather not get into these changes :p

ravicious replied:
> We use `memo` because rendering the `shared/UnifiedResources components is expensive. It is visible when quickly switching between tabs (it's not smooth, [see the original PR](https://github.com/gravitational/teleport/pull/36741#issue-2084194439)).

Sooner or later we're going to have to address the performance though. I don't think what this component does warrants how slow it is to render.

gzdunek replied:
>Sooner or later we're going to have to address the performance though. I don't think what this component is supposed to do necessitates it being slow.

Yes, I remember when I was fixing it for the first time, I tried to optimize the shared component, but I had some problems with it.

gzdunek committed:
Rephrase JSDoc for `options.clear`

gzdunek committed:
Remove unnecessary curly braces

gzdunek commented on web/packages/shared/hooks/useInfiniteScroll/useKeyBasedPagination.ts:
```diff
@@ -209,12 +212,16 @@ type KeyBasedPagination<T> = {
+   * flooding the server with requests.
+   * @param options.clear - If true, works like `force` but also clears
+   * the state. Similarly, do not call it from `useInfiniteScroll` to avoid
+   * flooding the server with requests.
```
In the end, I rephrased the JSDoc for `options.clear`. `options.force` simply cancels the previous request, I think repeating this information in `options.clear` is more clear than referring to the other option.
